### PR TITLE
[Merged by Bors] - feat: `Fintype (Finpartition s)` for `s : Finset α`

### PR DESCRIPTION
**Pull request by Parcly-Taxel**
Co-authored-by: Bhavik Mehta <bhavikmehta8@gmail.com>
Co-authored-by: Yakov Pechersky <ffxen158@gmail.com>

### CODE REVIEW

**Review by YaelDillies — APPROVED** [review]
Nicely done!

maintainer merge

**Comment by b-mehta** [comment]
This looks like it was ported from code I wrote in the Xena discord, without credit.

File: Mathlib/Order/Partition/Finpartition.lean
Comment (eric-wieser) [inline Mathlib/Order/Partition/Finpartition.lean:445]: This looks exponentially bad in terms of runtime / term size. Can we use the list function that was posted on Zulip which finds all partitions instead?

Comment (Parcly-Taxel) [reply Mathlib/Order/Partition/Finpartition.lean:445]: Where?

Comment (eric-wieser) [reply Mathlib/Order/Partition/Finpartition.lean:445]: Edited above with a link

Comment (Parcly-Taxel) [reply Mathlib/Order/Partition/Finpartition.lean:445]: The code in the Zulip link does not count the partitions I want. It merely takes in an ordered list and outputs all ways to put separators into it _preserving order_ – and the count of this is $2^{n-1}$ for a length-n list, not the Bell numbers.

Comment (eric-wieser) [reply Mathlib/Order/Partition/Finpartition.lean:445]: Here's the right function:
```
private def insertPerm (x : α) (l : List (List α)) : List (List (List α)) :=
  match l with
  | [] => [[[x]]]
  | y :: ys => ((x :: y) :: ys) :: (insertPerm x ys).map (y :: ·)

/-- Partitions of `l` into sublists. -/
def sublistPartitions (l : List α) : List (List (List α)) :=
  match l with
  | [] => [[]]
  | x :: xs => (sublistPartitions xs).bind fun p =>
    insertPerm x p
```

Comment (eric-wieser) [reply Mathlib/Order/Partition/Finpartition.lean:445]: I've opened #16647 with this definition; I am happy for you to try and solve the sorries there if you want to!

Comment (b-mehta) [reply Mathlib/Order/Partition/Finpartition.lean]: I agree there's scope to improve the runtime of this definition, but let's get this current one merged while the details of the faster one are figured out? If I recall correctly, it's also difficult to show this function descends correctly to an operation on `Finset`s. Let's add a TODO in the file to your PR as well, so it's recorded.

Comment (eric-wieser) [reply Mathlib/Order/Partition/Finpartition.lean]: Agreed, a TODO is enough to unblock this PR.

Comment (Parcly-Taxel) [reply Mathlib/Order/Partition/Finpartition.lean]: Done.

**Comment by b-mehta** [comment]
Thanks!

bors merge

**Pull request closed without merging**
